File: 116_jump-game/jump-game.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/jump-game
@Language: Java
@Datetime: 16-05-28 04:17
*/

public class Solution {
    /**
     * @param A: A list of integers
     * @return: The boolean answer
     */
    public boolean canJump(int[] A) {
        // wirte your code here
        //greedy algorithm
        if (A == null || A.length == 0) {
            return true;
        }
        int farthest = A[0];
        for (int i = 1; i < A.length; i++) {
            if (i <= farthest && i + A[i] > farthest) {
                farthest = i + A[i];
            }
        }
        return farthest >= A.length - 1;
    }
}

File: 116_jump-game/jump-game.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/jump-game
@Language: Python
@Datetime: 16-05-28 04:12
'''

class Solution:
    # @param A, a list of integers
    # @return a boolean
    def canJump(self, A):
        # write your code here
        # greedy algorithm, keep finding the farthest index
        if A is None or len(A) == 0:
            return True
        farthest = A[0]
        for i in range(1, len(A)):
            if i <= farthest and i + A[i] > farthest:
                farthest = i + A[i]
        return farthest >= len(A) - 1
        